packages/sdk-core/src/proving/circom/proving-manager-thread.ts

```diff
@@ -94,67 +90,49 @@ export class CircomProvingManagerThread {
+      if (notes.length !== indices.length) {
         throw new Error(
-          `Input notes and indices size don't match notes count (${rawNotes.length}) indices count (${indices.length})`
+          `Input notes and indices size don't match notes count (${notes.length}) indices count (${indices.length})`
         );
       }
 
       // Set the leaf index on the jsNotes
-      for (let i = 0; i < jsNotes.length; i++) {
-        jsNotes[i].mutateIndex(indices[i].toString());
+      for (let i = 0; i < notes.length; i++) {
+        notes[i].mutateIndex(indices[i].toString());
       }
 
       // sort input utxos
-      jsNotes.sort((a, b) => Number(a.sourceChainId) - Number(b.sourceChainId));
+      notes.sort((a, b) => Number(a.note.sourceChainId) - Number(b.note.sourceChainId));
 
-      let mt: MerkleTree = new MerkleTree(30, input.leavesMap[(jsNotes[0].sourceChainId)].map((u8a) => u8aToHex(u8a)));
+      // Account for empty leaves set on the merkle tree
+      let mt: MerkleTree = new MerkleTree(5, input.leavesMap[notes[0].note.sourceChainId]
+        ? input.leavesMap[(notes[0].note.sourceChainId)].map((u8a) => u8aToHex(u8a))
+        : []);
       const merkleProofs: MerkleProof[] = [];
 
-      merkleProofs.push(mt.path(Number(jsNotes[0].index)));
+      merkleProofs.push(mt.path(Number(notes[0].note.index)));
 
       // loop through the jsNotes and generate merkle proofs
-      for (let i = 1; i < jsNotes.length; i++) {
-        mt = new MerkleTree(30, input.leavesMap[(jsNotes[i].sourceChainId)].map((u8a) => u8aToHex(u8a)));
-
-        // generate the merkle proof for this note
-        merkleProofs.push(mt.path(Number(jsNotes[i].index)));
-      }
-
-      // Add extra dummy utxos to fit the circuit
-      if (rawNotes.length === 1) {
-        jsNotes.push(jsNotes[0].defaultUtxoNote());
-        indices.push(0);
-        merkleProofs.push({
-          element: BigNumber.from(0),
-          merkleRoot: BigNumber.from(mt.root()),
-          pathElements: new Array(30).fill(0),
-          pathIndices: new Array(30).fill(0)
-        });
-      }
-
-      if (rawNotes.length > 2 && rawNotes.length < 16) {
-        const inputGap = 16 - rawNotes.length;
-
-        jsNotes.push(
-          ...Array(inputGap)
-            .fill(0)
-            .map(() => jsNotes[0].defaultUtxoNote())
-        );
-        indices.push(...Array(inputGap).fill(0));
-        merkleProofs.push({
-          element: BigNumber.from(0),
-          merkleRoot: BigNumber.from(mt.root()),
-          pathElements: new Array(30).fill(0),
-          pathIndices: new Array(30).fill(0)
-        });
+      for (let i = 0; i < notes.length; i++) {
+        // generate the merkle proof for this note. If it is a dummy utxo, return zeros
+        if (notes[i].note.amount === '0') {
+          merkleProofs.push({
+            element: BigNumber.from(0),
+            merkleRoot: BigNumber.from(u8aToHex(input.roots[i])),
+            pathElements: new Array(5).fill(0),
```

Comment:
We should consider removing this hardcoded value and having it be provided by the function. What do you think?